### PR TITLE
Flexible memberwise initialization

### PR DESCRIPTION
**Pull request by anandabits**

### CODE REVIEW

File: proposals/NNNN-flexible-memberwise-initialization.md [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+
+1. Sometimes properties that should be immutable are made mutable and a potentially unsafe ad-hoc two-phase initialization pattern is employed where an instance is initialized and then configured immediately afterwards.  This allows the developer to avoid including boilerplate in every initializer that would otherwise be required to initialize immutable properties.
+
+2. Sometimes mutable properties that have a sensible default value are simply default-initialized and the same post-initialization configuration strategy is employed when the default value is not correct for the intended use.  This results in an instance which may pass through several states that are incorrect *for the intended use* before it is correctly initialized for its intended use.
+
+Underlying this problem is the fact initialization scales with M x N complexity (M members, N initializers).  We need as much help from the compiler as we can get!
```
Comment (beltex) [inline proposals/NNNN-flexible-memberwise-initialization.md]: **Minor**: Shouldn't this be the following?

> “_Underlying this problem is the fact **that** initialization scales..._”

**Comment by lattner** [comment]
Looks good, please squash it into one patch and I'll commit it!

**Comment by anandabits** [comment]
Thanks Chris.  I believe I have squashed them properly now!

I made a few further refinements to the interaction with access control you might want to review:
1. Clarify eligibility of properties with distinct setter access control.
2. Refine access control of implicitly generated memberwise initializer.
3. Add future enhancement to allow distinct access control for memberwise initialization eligibility.

**Comment by lattner** [comment]
Awesome!  We'll open the discussion period early next year, most likely the week of Jan 4th.  Thanks!

**Pull request merged**